### DISCUSSION TITLE
System hang when connecting to client that doesn't exist.

### DISCUSSION THREAD

**Discussion by videojedi** [ssilverman/QNEthernet#4] [General]
Hi, 
Firstly, many thanks for your hardwork, it's really helping me with a project.

However, I'm a little stuck with a client connection to a server that isn't there. My teensy just hangs with the following message

`Assertion "tcp_input: pcb->next != pcb (before cache)" failed at line 269 in C:\Users\admin\Documents\Arduino\libraries\QNEthernet\src\lwip\tcp_in.c
`

I'm just just initiating the connection with

`if (client.connect(setupIP.rtrIPaddr, setupIP.rtrPort)) {
blah
}
else {
Serial.println("connection failed!");
}`

Even the watchdog, is unable to recover the system hang.

Any thoughts greatly appreciated..

**Comment by ssilverman** [comment]
Thank you for the report. I will look when I have a chance. May I request a minimal full program that can reproduce this issue?

**Reply by ssilverman** [reply]
Also, would you mind creating an issue instead?

**Reply by videojedi** [reply]
I can of course, but there is a very good chance it's not an issue, and just my coding. :)

**Reply by ssilverman** [reply]
While it may have something to do with your coding, even if the coding is incorrect, it still means my library can expose an assert-level condition, which is a problem. I’ll have a look at your code…

**Comment by ssilverman** [comment]
I tried duplicating the error, but could not. I tried connecting to several valid addresses, local and global, several addresses I know can’t connect, local and global, then repeated all the tests without first calling `Ethernet.begin()`. I connected to port 80 and did a rudimentary HTTP connection. All my results were expected and I didn’t see any lwip asserts.

Are you, by any chance, trying to use the library in a multithreaded context? _QNEthernet_ isn’t thread-safe.

I’ll need some code that generates the issue.

**Comment by videojedi** [comment]
I'm just trying to unpick my code, so I can send you something useful. (Also tidy it up :))

The teensy is hosting a webpage (EthernetWebServer.h) and is using Websockets (WebSockets2_Generic.h) to communicate between the browser and Teensy. Then I'm using a client connection to another piece of hardware. It all works amazingly well, until I use the web gui to change the IP address of the hardware I'm connecting to that doesn't exist. The crash seems to co-incide with a failed connection to hardware and a web page refresh or websocket update request.

Thanks for checking.

**Reply by videojedi** [reply]
Apologies, I've stripped out as much as I can, but leaving the barebones of the web gui. If you comment out line 157, then the webpage should load and you can enter the IP address of target device.

`    //comment following line to allow web gui to run again after crash
    PollHardware(); `

https://github.com/videojedi/js3debug.git

**Reply by ssilverman** [reply]
Thanks! I’ll have a look when I can.

**Comment by ssilverman** [comment]
I just pushed some changes that _may_ fix the issue, specifically commit 7724fbe68ace503e2e239546c9b9c935f9cd2f1c. See the latest `master`.

**Comment by ssilverman** [comment]
See also this commit: b69228f5a72ebf960f04f4d47d620b3741c47387

**Comment by ssilverman** [comment]
Closing for now. Please reopen if you're still seeing this.
